src/backend/distributed/commands/index.c

```diff
@@ -172,65 +188,108 @@ PlanIndexStmt(IndexStmt *createIndexStatement, const char *createIndexCommand)
+ * PlanReindexStmt determines whether a given REINDEX statement involves
+ * a distributed table. If so (and if the statement does not use unsupported
+ * options), it modifies the input statement to ensure proper execution against
+ * the master node table and creates a DDLJob to encapsulate information needed
+ * during the worker node portion of DDL execution before returning that DDLJob
+ * in a List. If no distributed table is involved, this function returns NIL.
  */
-void
-ErrorIfReindexOnDistributedTable(ReindexStmt *ReindexStatement)
+List *
+PlanReindexStmt(ReindexStmt *ReindexStatement, const char *ReindexCommand)
 {
-	Relation relation = NULL;
-	Oid relationId = InvalidOid;
-	bool isDistributedRelation = false;
-	LOCKMODE lockmode = AccessShareLock;
+	List *ddlJobs = NIL;
 
 	/*
 	 * We first check whether a distributed relation is affected. For that, we need to
-	 * open the relation.
+	 * open the relation. To prevent race conditions with later lookups, lock the table,
+	 * and modify the rangevar to include the schema.
 	 */
-	if (ReindexStatement->relation == NULL)
+	if (ReindexStatement->relation != NULL)
 	{
-		/* ignore REINDEX SCHEMA, REINDEX SYSTEM, and REINDEX DATABASE */
-		return;
-	}
+		Relation relation = NULL;
+		Oid relationId = InvalidOid;
+		bool isDistributedRelation = false;
+		char *namespaceName = NULL;
+#if PG_VERSION_NUM >= 120000
+		LOCKMODE lockmode = ReindexStatement->concurrent ? ShareUpdateExclusiveLock :
+							AccessExclusiveLock;
+#else
+		LOCKMODE lockmode = AccessExclusiveLock;
+#endif
+		MemoryContext relationContext = NULL;
 
-	Assert(ReindexStatement->kind == REINDEX_OBJECT_INDEX ||
-		   ReindexStatement->kind == REINDEX_OBJECT_TABLE);
+		Assert(ReindexStatement->kind == REINDEX_OBJECT_INDEX ||
+			   ReindexStatement->kind == REINDEX_OBJECT_TABLE);
 
-	/*
-	 * XXX: Consider using RangeVarGetRelidExtended with a permission
-	 * checking callback. Right now we'll acquire the lock before having
-	 * checked permissions.
-	 */
-	if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
-	{
-		Oid indOid = RangeVarGetRelid(ReindexStatement->relation,
-									  NoLock, false);
-		relation = index_open(indOid, lockmode);
-		relationId = IndexGetRelation(indOid, false);
-	}
-	else
-	{
-		relation = heap_openrv(ReindexStatement->relation, lockmode);
-		relationId = RelationGetRelid(relation);
-	}
+		if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
+		{
+			Oid indOid;
+			struct ReindexIndexCallbackState state;
+#if PG_VERSION_NUM >= 120000
+			state.concurrent = ReindexStatement->concurrent;
+#endif
+			state.locked_table_oid = InvalidOid;
+
+			indOid = RangeVarGetRelidInternal(ReindexStatement->relation,
+											  lockmode, 0,
+											  RangeVarCallbackForReindexIndex,
+											  &state);
+			relation = index_open(indOid, NoLock);
+			relationId = IndexGetRelation(indOid, false);
+		}
+		else
+		{
+			Oid tblOid = RangeVarGetRelid(ReindexStatement->relation,
+										  NoLock, false);
```

Comment:
I suppose this should be
```
   tblOid = RangeVarGetRelidExtended(relation, ShareLock, 0,
                                       RangeVarCallbackOwnsTable, NULL);
```

This one should be simpler since `RangeVarCallbackOwnsTable` is a public function in `commands/tablecmds.h`.

Reply:
Had checked that `EnsureTableOwner` achieves the same thing, but updated to use `RangeVarCallbackOwnsTable` instead